source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -877,8 +889,19 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
         }
       }
       if (!downstream_closed_) {
-        route_ = pickRoute();
-        establishUpstreamConnection();
+        if (connect_attempts_ >= config_->maxConnectAttempts()) {
```

Comment:
The default `maxConnectAttempt` is 1, and here we put the check after the `connect_attemts_++` in `maybeTunnel()`, we will always hit this since `1 >= 1`, you maybe need to adjust the counter check here. Could you also a test case to very this, please?

Reply:
nvm, I think this makes sense now, we put the check forward, and this condition will trigger when the connection attempt fails and right before the new attempt. And it could be cleaner to move this block to a retry handler as a whole, but not a strong opinion.